File: core/image/json_importer.py
"""
JSON 导入器 v2.3.2
解析第三方像素画 JSON 文件（如 living-the-grid 格式），转换为内部格式。
支持顺滑画笔和像素画笔两种模式，处理网格居中与缩放。
适配新版 JSON 中的 press 对象（h/s/b 对应 ZR/方向键右/方向键上按下次数），
并处理 h=200 需替换为 199、h=201 需替换为 0 的特殊规则。
颜色匹配使用 HEX 字符串。
顺滑画笔生效区域为居中扩展（基于游戏底层 a=(n-1)/2, b=(n+1)/2 公式）。
"""

import json
import numpy as np
from typing import List, Tuple, Dict, Any, Optional

from core.utils.logger import get_logger
from core.image.preset_palette import get_preset_palette_hex
from core.models.canvas_mode import CanvasMode, get_canvas_mode, json_preset_to_canvas_mode, CANVAS_MODE_DISPLAY


class JsonImporter:
    """JSON 像素画导入器"""

    def __init__(self):
        self.logger = get_logger("JsonImporter")

    def load_from_file(
        self,
        file_path: str,
        brush_type: str,
        brush_size: int,
        canvas_mode: str = "standard",
    ) -> Tuple[Optional[np.ndarray], Optional[List[List[int]]], Dict[str, Any]]:
        """
        从 JSON 文件加载像素画数据。

        Args:
            file_path: JSON 文件路径
            brush_type: 画笔类型，"smooth"（顺滑）或 "pixel"（像素）
            brush_size: 画笔尺寸（1,3,7,13,19,27 或 4,8,16,32）
            canvas_mode: 画布模式（"standard", "book", "tv", "game", "decoration"）

        Returns:
            (color_index_matrix, color_palette, metadata)
            - color_index_matrix: 256×256 颜色索引矩阵（-1 表示透明或无效区域）
            - color_palette: RGB 调色板列表 [[r,g,b], ...]
            - metadata: 包含 width, height, brush_type, brush_size 等信息的字典
        """
        mode = get_canvas_mode(canvas_mode)
        self.logger.info(f"开始导入 JSON 文件: {file_path}")
        self.logger.info(f"画笔类型: {brush_type}, 尺寸: {brush_size}, 画布模式: {mode.name}")

        # 1. 读取并解析 JSON
        try:
            with open(file_path, "r", encoding="utf-8") as f:
                data = json.load(f)
        except Exception as e:
            self.logger.error(f"读取 JSON 文件失败: {e}")
            return None, None, {"error": f"读取文件失败: {e}"}

        # 2. 校验必要字段
        required_fields = ["width", "height", "palette", "pixels"]
        for field in required_fields:
            if field not in data:
                self.logger.error(f"JSON 缺少必要字段: {field}")
                return None, None, {"error": f"JSON 格式错误，缺少字段: {field}"}

        width = data["width"]
        height = data["height"]
        raw_palette = data["palette"]
        raw_pixels = data["pixels"]

        # 3. 验证调色板与像素数组格式
        if not isinstance(raw_palette, list) or not isinstance(raw_pixels, list):
            return None, None, {"error": "调色板或像素数组格式错误"}

        if len(raw_pixels) != height:
            self.logger.warning(
                f"像素行数 ({len(raw_pixels)}) 与高度 ({height}) 不匹配"
            )
            height = min(height, len(raw_pixels))

        # 4. 转换调色板：十六进制 -> RGB 元组列表，同时提取 press 数据
        color_palette = []  # 返回的 RGB 列表
        hex_palette = []  # 内部用于与预设表比较的 HEX 列表
        press_data = None
        has_press = False

        for item in raw_palette:
            if isinstance(item, dict):
                # 新版 JSON 格式：{"hex": "#...", "rgb": [...], "press": {...}}
                hex_str = item.get("hex", "")
                if not hex_str.startswith("#"):
                    return None, None, {"error": f"无效的调色板颜色: {item}"}
                hex_str_clean = hex_str.lstrip("#")
                try:
                    r = int(hex_str_clean[0:2], 16)
                    g = int(hex_str_clean[2:4], 16)
                    b = int(hex_str_clean[4:6], 16)
                    color_palette.append([r, g, b])
                except Exception:
                    return None, None, {"error": f"解析颜色失败: {item}"}

                # 保存原始 HEX（统一大写）
                hex_palette.append(hex_str.upper())
                has_press = "press" in item

            elif isinstance(item, str):
                # 旧版格式：纯十六进制字符串，无 press
                hex_str = item.lstrip("#")
                try:
                    r = int(hex_str[0:2], 16)
                    g = int(hex_str[2:4], 16)
                    b = int(hex_str[4:6], 16)
                    color_palette.append([r, g, b])
                except Exception:
                    return None, None, {"error": f"解析颜色失败: {item}"}
                hex_palette.append(item.upper())
            else:
                return None, None, {"error": f"无效的调色板颜色: {item}"}

        # 统一处理 press 数据（新版字典格式）
        if has_press:
            press_data = []
            for item in raw_palette:
                if isinstance(item, dict) and "press" in item:
                    p = item["press"]
                    h = p.get("h", 0)
                    if h == 200:
                        h = 199
                    elif h == 201:
                        h = 0
                    s = p.get("s", 0)
                    b = p.get("b", 0)
                    press_data.append({"h": h, "s": s, "b": b})
                else:
                    press_data.append({"h": 0, "s": 0, "b": 0})

        # 5. 验证所有颜色是否在预设84色内（基于 HEX 判断）
        preset_hex_set = set(get_preset_palette_hex())
        missing_hex = [h for h in hex_palette if h not in preset_hex_set]

        if missing_hex:
            self.logger.warning(
                f"调色板中有 {len(missing_hex)} 种颜色不在预设84色内，将使用自定义模式"
            )
        else:
            self.logger.info("所有颜色均在预设84色内，可使用预设模式绘图")

        # 6. 计算网格缩放与居中偏移
        cell_span = brush_size
        pattern_width = width * cell_span
        pattern_height = height * cell_span

        if pattern_width > 256 or pattern_height > 256:
            self.logger.warning(
                f"图案尺寸 ({pattern_width}x{pattern_height}) 超过 256x256，将被裁剪"
            )

        offset_x = (256 - pattern_width) // 2
        offset_y = (256 - pattern_height) // 2

        # 7. 构建 256x256 颜色索引矩阵
        color_index_matrix = np.full((256, 256), -1, dtype=np.int16)

        for row in range(height):
            if row >= len(raw_pixels):
                break
            row_data = raw_pixels[row]
            if not isinstance(row_data, list):
                continue
            for col in range(min(width, len(row_data))):
                idx = row_data[col]
                if idx is None:
                    continue
                if not isinstance(idx, int) or idx < 0 or idx >= len(color_palette):
                    continue

                base_x = offset_x + col * cell_span
                base_y = offset_y + row * cell_span

                # 顺滑画笔居中填充：从 (cx-a, cy-a) 开始填 n×n
                if brush_type == "smooth" and cell_span > 1:
                    a = (cell_span - 1) // 2
                    cx = base_x + cell_span // 2
                    cy = base_y + cell_span // 2
                    fill_start_x = cx - a
                    fill_start_y = cy - a
                else:
                    fill_start_x = base_x
                    fill_start_y = base_y

                for dy in range(cell_span):
                    y = fill_start_y + dy
                    if y < 0 or y >= 256:
                        continue
                    for dx in range(cell_span):
                        x = fill_start_x + dx
                        if x < 0 or x >= 256:
                            continue
                        color_index_matrix[y, x] = idx

        # 解析 JSON 顶层 canvas 字段（画布模式）
        json_canvas = data.get("canvas")
        if isinstance(json_canvas, dict):
            preset = json_canvas.get("preset")
            if preset:
                detected_mode = json_preset_to_canvas_mode(preset)
                mode_obj = get_canvas_mode(detected_mode)
                cw = json_canvas.get("w")
                ch = json_canvas.get("h")
                if cw is not None and ch is not None:
                    if cw != mode_obj.active_w or ch != mode_obj.active_h:
                        self.logger.warning(
                            f"JSON canvas 尺寸 ({cw}x{ch}) 与预设 {preset} "
                            f"({mode_obj.active_w}x{mode_obj.active_h}) 不匹配，以预设为准"
                        )
                self.logger.info(
                    f"检测到画布模式: {preset} → {CANVAS_MODE_DISPLAY[detected_mode]}"
                )
            else:
                detected_mode = canvas_mode
        else:
            detected_mode = canvas_mode

        # 8. 解析顶层 brush 字段（如果存在）
        json_brush_type = None
        json_brush_size = None
        brush_data = data.get("brush")
        if isinstance(brush_data, dict):
            bmode = brush_data.get("mode")
            px = brush_data.get("px")
            if bmode in ("smooth", "pixel"):
                json_brush_type = bmode
            else:
                self.logger.warning(f"未知的画笔模式: {bmode}，忽略")
            if isinstance(px, int):
                json_brush_size = px
            else:
                if json_brush_type is not None:
                    self.logger.warning(f"无效的画笔尺寸: {px}")
                    json_brush_type = None

        # 9. 构建元数据
        metadata = {
            "source": data.get("source", ""),
            "width": width,
            "height": height,
            "brush_type": brush_type,
            "brush_size": brush_size,
            "palette_size": len(color_palette),
            "total_pixels": np.sum(color_index_matrix >= 0),
            "offset": (offset_x, offset_y),
            "all_preset": len(missing_hex) == 0,
            "json_brush_type": json_brush_type,
            "json_brush_size": json_brush_size,
            "canvas_mode": detected_mode,
        }
        if press_data is not None:
            metadata["press_data"] = press_data

        self.logger.info(f"导入完成，有效像素数: {metadata['total_pixels']}")
        return color_index_matrix, color_palette, metadata

File: frontend/src/engine/json-importer.ts
import { getPresetPaletteHex } from './preset-palette';
import { jsonPresetToCanvasMode, CANVAS_MODE_DISPLAY } from './canvas-mode';
import { Point } from './types';

export interface ImportResult {
  /** 256×256 颜色索引矩阵 (-1 = transparent) */
  matrix: Int16Array;
  /** 颜色索引波形 */
  matrix2d: number[][];
  /** RGB调色板 */
  palette: [number, number, number][];
  /** 元数据 */
  metadata: ImportMeta;
}

export interface ImportMeta {
  source: string;
  width: number;
  height: number;
  brushType: string;
  brushSize: number;
  paletteSize: number;
  totalPixels: number;
  offsetX: number;
  offsetY: number;
  allPreset: boolean;
  canvasMode: string;
  jsonBrushType?: string;
  jsonBrushSize?: number;
  pressData?: { h: number; s: number; b: number }[];
  error?: string;
}

/**
 * 解析第三方 living-the-grid JSON 文件。
 * 支持 preset / custom 两种调色盘，顺滑 / 像素两种画笔。
 */
export function importJson(
  data: any,
  brushType: string,
  brushSize: number,
  canvasMode: string,
): ImportResult {
  // 校验
  if (!data || !data.width || !data.height || !data.palette || !data.pixels) {
    return errorResult({ error: 'JSON 格式错误，缺少必要字段' });
  }

  const width: number = data.width;
  const height: number = Math.min(data.height, data.pixels.length);
  const rawPalette: any[] = data.palette;
  const rawPixels: (number | null)[][] = data.pixels;

  // 解析调色板
  const colorPalette: [number, number, number][] = [];
  const hexPalette: string[] = [];
  let pressData: { h: number; s: number; b: number }[] | undefined;
  let hasPress = false;

  for (const item of rawPalette) {
    if (typeof item === 'object' && item.hex) {
      const hex = item.hex.replace('#', '');
      colorPalette.push([
        parseInt(hex.substring(0, 2), 16),
        parseInt(hex.substring(2, 4), 16),
        parseInt(hex.substring(4, 6), 16),
      ]);
      hexPalette.push(item.hex.toUpperCase());
      hasPress = !!item.press;
    } else if (typeof item === 'string') {
      const hex = item.replace('#', '');
      colorPalette.push([
        parseInt(hex.substring(0, 2), 16),
        parseInt(hex.substring(2, 4), 16),
        parseInt(hex.substring(4, 6), 16),
      ]);
      hexPalette.push(item.toUpperCase());
    } else {
      return errorResult({ error: `无效的调色板颜色: ${item}` });
    }
  }

  if (hasPress) {
    pressData = rawPalette.map(item => {
      if (item.press) {
        let h = item.press.h || 0;
        if (h === 200) h = 199;
        else if (h === 201) h = 0;
        return { h, s: item.press.s || 0, b: item.press.b || 0 };
      }
      return { h: 0, s: 0, b: 0 };
    });
  }

  // 验证预设色
  const presetHexSet = new Set(getPresetPaletteHex());
  const missingHex = hexPalette.filter(h => !presetHexSet.has(h));
  const allPreset = missingHex.length === 0;

  // 计算居中偏移
  const cellSpan = brushSize;
  const patternW = width * cellSpan;
  const patternH = height * cellSpan;
  const offsetX = (256 - patternW) >> 1;
  const offsetY = (256 - patternH) >> 1;

  // 构建 256×256 矩阵
  const size = 256 * 256;
  const matrix = new Int16Array(size);
  matrix.fill(-1);

  for (let row = 0; row < height; row++) {
    if (row >= rawPixels.length) break;
    const rowData = rawPixels[row];
    if (!Array.isArray(rowData)) continue;
    for (let col = 0; col < Math.min(width, rowData.length); col++) {
      const idx = rowData[col];
      if (idx === null || idx === undefined) continue;
      if (typeof idx !== 'number' || idx < 0 || idx >= colorPalette.length) continue;

      const baseX = offsetX + col * cellSpan;
      const baseY = offsetY + row * cellSpan;

      let fillStartX: number, fillStartY: number;
      if (brushType === 'smooth' && cellSpan > 1) {
        const a = (cellSpan - 1) >> 1;
        const cx = baseX + (cellSpan >> 1);
        const cy = baseY + (cellSpan >> 1);
        fillStartX = cx - a;
        fillStartY = cy - a;
      } else {
        fillStartX = baseX;
        fillStartY = baseY;
      }

      for (let dy = 0; dy < cellSpan; dy++) {
        const y = fillStartY + dy;
        if (y < 0 || y >= 256) continue;
        for (let dx = 0; dx < cellSpan; dx++) {
          const x = fillStartX + dx;
          if (x < 0 || x >= 256) continue;
          matrix[y * 256 + x] = idx;
        }
      }
    }
  }

  // 转换为 2D 数组
  const matrix2d: number[][] = [];
  for (let y = 0; y < 256; y++) {
    const row: number[] = [];
    for (let x = 0; x < 256; x++) {
      row.push(matrix[y * 256 + x]);
    }
    matrix2d.push(row);
  }

  // 解析 JSON canvas mode
  let detectedMode = canvasMode;
  const jsonCanvas = data.canvas;
  if (jsonCanvas?.preset) {
    detectedMode = jsonPresetToCanvasMode(jsonCanvas.preset);
  }

  // 解析 JSON brush
  let jsonBrushType: string | undefined;
  let jsonBrushSize: number | undefined;
  if (data.brush) {
    if (data.brush.mode === 'smooth' || data.brush.mode === 'pixel') {
      jsonBrushType = data.brush.mode;
    }
    if (typeof data.brush.px === 'number') {
      jsonBrushSize = data.brush.px;
    }
  }

  let totalPixels = 0;
  for (let i = 0; i < size; i++) {
    if (matrix[i] >= 0) totalPixels++;
  }

  return {
    matrix, matrix2d, palette: colorPalette,
    metadata: {
      source: data.source || '',
      width, height, brushType, brushSize,
      paletteSize: colorPalette.length, totalPixels,
      offsetX, offsetY, allPreset,
      canvasMode: detectedMode,
      jsonBrushType, jsonBrushSize,
      pressData,
    },
  };
}

function errorResult(meta: Partial<ImportMeta>): ImportResult {
  const size = 256 * 256;
  const matrix = new Int16Array(size);
  matrix.fill(-1);
  return {
    matrix,
    matrix2d: [],
    palette: [],
    metadata: {
      source: '', width: 0, height: 0,
      brushType: 'smooth', brushSize: 1,
      paletteSize: 0, totalPixels: 0,
      offsetX: 0, offsetY: 0, allPreset: false,
      canvasMode: 'standard',
      ...meta,
    },
  };
}
